src/translation.cpp

```diff
@@ -0,0 +1,760 @@
+/*
+ * This file is part of EasyRPG Player.
+ *
+ * EasyRPG Player is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * EasyRPG Player is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with EasyRPG Player. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "translation.h"
+
+// Headers
+#include <fstream>
+#include <iomanip>
+#include <lcf/data.h>
+#include <lcf/rpg/terms.h>
+#include <lcf/rpg/map.h>
+#include "lcf/rpg/mapinfo.h"
+
+#include "cache.h"
+#include "main_data.h"
+#include "game_actors.h"
+#include "game_map.h"
+#include "player.h"
+#include "output.h"
+#include "utils.h"
+
+
+// Name of the translate directory
+#define TRDIR_NAME "languages"
+
+// Name of expected files
+#define TRFILE_RPG_RT_LDB    "rpg_rt.ldb.po"
+#define TRFILE_RPG_RT_COMMON "rpg_rt.ldb.common.po"
+#define TRFILE_RPG_RT_BATTLE "rpg_rt.ldb.battle.po"
+#define TRFILE_RPG_RT_LMT    "rpg_rt.lmt.po"
+#define TRFILE_META_INI      "META.INI"
+
+// Message box commands to remove a message box or add one in place.
+#define TRCUST_REMOVEMSG        "__EASY_RPG_CMD:REMOVE_MSGBOX__"
+#define TRCUST_ADDMSG           "__EASY_RPG_CMD:ADD_MSGBOX__"
```

Comment:
Applies to remove and add: Is this still used? I see them checked in ifs but never assigned?

Reply:
These strings are supposed to be added by the translator. Here's how it works; let's say I have a message box:

```
msgid ""
"Hero: Hello there!"
```

...and I want that message box to be removed from the translation entirely. In that case, I replace it with:

```
msgstr ""
"__EASY_RPG_CMD:REMOVE_MSGBOX__"
```

Let's say, however, that in the target language it makes more sense for the other party to shout at the hero first. So I need a new message box to appear BEFORE the current one. I'll replace it with:

```
msgstr ""
"Townsperson: Mr. Hero, hello!"
"__EASY_RPG_CMD:ADD_MSGBOX__"
"Hero: Hello there!"
```

This will ADD a message box to the event list, so now we'll have two message boxes (rather than a single two-line message box).

They are magic strings, unfortunately, but I think they're pretty unlikely to appear in source games. (And I can't think of a more robust way to do this that interacts cleanly with gettext).

Reply:
ah now I get it so that's some kind of in-band signaling. Will see if I can find a way to improve this.

For "insert new page" we already have the the form feed "\f" (needed for some battle messages) but I guess this will need changes to the po parser.

For "delete message" one could use gettext flags. They are marked via ``#,``, something like ``#, easyrpg-delete`` (see https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html). Will also need parser changes (to ensure lcftrans keeps them)

Reply:
Yep, that's right. 

I think I *might* prefer to keep the current strings, since they are very self-descriptive for translators, and I don't think that "\f" or gettext flags will clean up our own processing code. But this is certainly up for debate.

Reply:
at least \f will not easily work because our code filters away control characters... Have to sleep about this ^^'